spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,426 @@
+package org.apache.spark.sql.cassandra
+
+import scala.reflect.ClassTag
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.catalyst.expressions.{Literal, AttributeReference, Expression, Attribute}
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.{ReadConf, ValidRDDType}
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+/**
+ * Table scan type. It allows to create a Relation for given table, keyspace, cluster and user
+ * defined schema. It also allow to pass [[CassandraConnector]], [[ReadConf]] and [[WriteConf]]
+ * explicitly
+ */
+sealed trait ScanType {
```

Comment:
What performance gain are you talking about? Can you elaborate on this? 

CatalystScan is an experimental API that allows to work on raw expressions, instead of a portable Filter implementation, which is simpler, but more limited. But you are *not* taking advantage of any things provided by `Expression` over `Filter`, so there is no perfomance advantage. Both scans hook into exactly the same place in Spark code-base.